### PR TITLE
test: Support for Multi-Level Partition Tables

### LINKED ISSUE #56: Support multi-level partition tables
### What feature are you requesting?

Support for foreign tables in a multi-level partitioned table setup (see example).

### Why are you requesting this feature?

When you have a lot of large files in one object store key/directory, scanning all of the files can be prohibitively time consuming. Changing the storage structure to support hive keys may not always be an option.

### What is your proposed implementation for this feature?

This is an example of a multi-level partition table, but in many cases only the first partition level would be needed.

Example:

A “root” table that will have two partition “levels”:

```
CREATE TABLE my_table(
  <column definitions>
) PARTITION BY list(id_1);
```

The table for the first partition level:

```
create table if not exists my_table_123
partition of my_table
for values in ('123') partition by list(id_2);
```

Many partition tables may be created at this level - one for each `id_1` value in the root table.

Note: If these first two tables have to be created as foreign tables in order for the last partition level table to be a foreign table, they shouldn’t require a files option.

The table for the next partition level looks like this:

```
create foreign table if not exists my_table_123_abc
partition of my_table_123
for values in ('abc')
options (files 's3://<bucket_name>/.../<filename>.parquet');
```

Many partition tables may be created at this level as well - one for each `id_2` value in the first partition level tables.

Note: when creating this final “leaf” partition table as a foreign table, a `files` option should be required.

So, when querying the root table with a where clause like this:

```
select *
from my_table
where id_1 = '123'
and   id_2 = 'abc'
```

...only one parquet file should be scanned - the one specified by the files option of the “leaf” partition table. Or, if a glob pattern is used in the `files` option of the leaf partition table, only the (ideally) few number of matching files will be scanned.

### Full Name:

Patrick Park

### Affiliation:

Payzer

### PR DESCRIPTION
**Pull request by shamb0**
Closes #56 

## What

Implements a demonstration test for multi-level partition tables, addressing issue [#56](https://github.com/paradedb/pg_analytics/issues/56).

## Why

This demo showcases the `pg_analytics` extension's capability to support multi-level partitioned tables. The implementation organizes data hierarchically, enabling efficient access to context-relevant information.

## How

1. Generates a simulated Auto Sales dataset and creates a local Parquet data source.
2. Utilizes DataFusion to load sales data from the Parquet file into a DataFrame.
3. Partitions the data and uploads partitions to an S3 bucket.
4. Configures necessary tables and `pg_analytics` Foreign Data Wrapper (FDW) in PostgreSQL using S3 data.
5. Executes analytics test queries on the multi-level partitioned table.

## Tests

To run the demonstration test:

```
RUST_LOG=info \
    cargo test \
    --test \
    test_mlp_auto_sales \
    -- \
    --nocapture
```

**Test traces are available in the attached log file ::** https://gist.githubusercontent.com/shamb0/2ed909ac9604c610af1d7fa0e87f9a82/raw/02a4203cdc1d675181d9f9700c578c81405becdb/wk2434-pg_analytics-mlp-demo.md

### CODE REVIEW

**Review by rebasedming — COMMENTED** [review]
Could you explain in the PR README how you set up the Postgres tables to match the Parquet partitions? It looks like this is something we'll need to document in the main repo to show users how to do.

**Comment by shamb0** [comment]
> Could you explain in the PR README how you set up the Postgres tables to match the Parquet partitions? It looks like this is something we'll need to document in the main repo to show users how to do.

Hi @rebasedming,

I’ve put together a detailed README for the PR, which you can find [here](https://gist.github.com/shamb0/1f946ed03ef2890029971036fb9b2db4).

I aimed to be thorough in capturing all the necessary details. If you find it too lengthy or in need of restructuring, please let me know, and I’ll make the necessary adjustments.

File: tests/datasets/auto_sales/mod.rs [hunk 1]
```diff
@@ -0,0 +1,718 @@
...
+            ],
+            "Pg_Analytics".to_string(),
+            &total_sales_results,
+            "DataFrame".to_string(),
+            &expected_results,
+        )
+        .await?;
+
+        // Compare the results with a small epsilon for floating-point precision.
+        let epsilon = 0.0001;
```
Comment (rebasedming) [inline tests/datasets/auto_sales/mod.rs:478]: You can use `assert_relative_eq` for this

File: tests/datasets/auto_sales/mod.rs [hunk 2]
```diff
@@ -0,0 +1,718 @@
...
+            assert_eq!(pg_year, df_year, "Year mismatch");
+            assert_eq!(pg_manufacturer, df_manufacturer, "Manufacturer mismatch");
+            assert!(
+                (pg_total - df_total).abs() < epsilon,
+                "Total sales for year {} and manufacturer {} differ more than expected epsilon: PostgreSQL = {}, DataFrame = {}",
+                pg_year, pg_manufacturer, pg_total, df_total
+            );
+        }
+
+        tracing::info!("Time taken for Pg_Analytics query: {:?}", pg_duration);
```
Comment (rebasedming) [inline tests/datasets/auto_sales/mod.rs:491]: Tests should just be for asserting results. Performance times should be measured in separate benchmarks, let's remove this.

**Review by rebasedming — COMMENTED** [review]
I've put up https://github.com/paradedb/pg_analytics/pull/91 which brings all our text fixtures into this crate. If possible, I'd appreciate it if you could align your PR with this one once it's merged. For instance, you can now put your fixtures as a new file in the `fixtures` crate, and `auto_sales` can go under `tables`.

File: tests/common/print_utils.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            fn to_row(&self) -> Vec<String> {
+                let ($($T,)+) = self;
+                vec![$($T.to_string(),)+]
+            }
+        }
+    }
+}
+
+// Implement Printable for tuples up to 12 elements
+impl_printable_for_tuple!(T1);
```
Comment (rebasedming) [inline tests/common/print_utils.rs:41]: Let's not bundle this into this PR, it's a separate thing that we can discuss whether it's worth bringing into the repo.

**Comment by rebasedming** [comment]
> > Could you explain in the PR README how you set up the Postgres tables to match the Parquet partitions? It looks like this is something we'll need to document in the main repo to show users how to do.
> 
> Hi @rebasedming,
> 
> I’ve put together a detailed README for the PR, which you can find [here](https://gist.github.com/shamb0/1f946ed03ef2890029971036fb9b2db4).
> 
> I aimed to be thorough in capturing all the necessary details. If you find it too lengthy or in need of restructuring, please let me know, and I’ll make the necessary adjustments.

Nice, thank you for the extremely thorough writeup.

What I was most interested in is your approach of putting partitioned heap tables in front of foreign tables to pass partition keys to the Parquet file string. I wasn't aware this was possible and was hoping you could elaborate on that.

**Comment by shamb0** [comment]
> tests/common/print_utils.rs

"Thanks for pointing out the missing critical piece. I've introduced the section [Partitioned Table Structure and S3 Integration](https://gist.github.com/shamb0/1f946ed03ef2890029971036fb9b2db4#partitioned-table-structure-and-s3-integration) with the necessary details, and I've also created a [TL;DR quick overview](https://gist.github.com/shamb0/ff50fba34822781464d8fdc15ce2f797)."

File: tests/datasets/auto_sales/mod.rs (hunk 1 above)
Comment (shamb0) [reply tests/datasets/auto_sales/mod.rs:478]: Sure, done.

File: tests/datasets/auto_sales/mod.rs (hunk 2 above)
Comment (shamb0) [reply tests/datasets/auto_sales/mod.rs:491]: Makes sense, I've removed the unnecessary parts.

File: tests/common/print_utils.rs (hunk 3 above)
Comment (shamb0) [reply tests/common/print_utils.rs:41]: Sure, done. I've removed the macros. Let me know if you want `tests/common/print_utils.rs` removed as well.

**Comment by shamb0** [comment]
Hi @rebasedming,

I've made some changes to address the `clippy` warnings in this PR:

1. Created a new crate: `pg_analytics_test_helpers`
2. Moved `fixtures` and `common` modules from the `tests` folder to this new crate

These changes were necessary to get the CI build passing. I know this might be outside the scope of our current PR, so let me know if you'd prefer I revert these changes.

Thanks for your input on this.

**Comment by philippemnoel** [comment]
> Hi @rebasedming,
> 
> I've made some changes to address the `clippy` warnings in this PR:
> 
> 1. Created a new crate: `pg_analytics_test_helpers`
> 2. Moved `fixtures` and `common` modules from the `tests` folder to this new crate
> 
> These changes were necessary to get the CI build passing. I know this might be outside the scope of our current PR, so let me know if you'd prefer I revert these changes.
> 
> Thanks for your input on this.

Can you explain which Clippy warnings you needed a new crate for? I don't think we should complicate the project with an extra crate. We can simply ignore some clippy warnings in the test files if they're causing issues.

**Comment by philippemnoel** [comment]
I also took a look at your CI error, and I'm a bit confused by it. I suspect this has to do with the newly introduced crate. If you keep a single crate, it shouldn't find two different PG versions.

**Comment by shamb0** [comment]
Hi @philippemnoel,

I've identified an issue with commit `9491b824e5dfffd3d636dcb7bb4b67a3fe9ee858`:

1. `cargo test` runs successfully
2. `cargo clippy --all-targets` fails

The problem:
- Error occurs in: `use crate::common::{execute_query, fetch_results, print_utils};`
- Clippy searches for the `common` module in the `pg_analytics` crate instead of the `tests` folder

I tried several solutions (e.g., using `super`) to fix the module path. But no luck. As a temporary fix, I've:
- Moved the module `fixture` and `common` out of the `tests` folder
- Created a new crate: `pg_analytics_test_helpers`

image

**Comment by philippemnoel** [comment]
> Hi @philippemnoel,
> 
> I've identified an issue with commit `9491b824e5dfffd3d636dcb7bb4b67a3fe9ee858`:
> 
> 1. `cargo test` runs successfully
> 2. `cargo clippy --all-targets` fails
> 
> The problem:
> 
> * Error occurs in: `use crate::common::{execute_query, fetch_results, print_utils};`
> * Clippy searches for the `common` module in the `pg_analytics` crate instead of the `tests` folder
> 
> I tried several solutions (e.g., using `super`) to fix the module path. But no luck. As a temporary fix, I've:
> 
> * Moved the module `fixture` and `common` out of the `tests` folder
> * Created a new crate: `pg_analytics_test_helpers`
> 
> image

This is probably an artifact of when we split this project from the monorepo. Can you simply fix the import path instead and keep it all in the same crate?

**Comment by philippemnoel** [comment]
> Hi @philippemnoel,
> 
> I've identified an issue with commit `9491b824e5dfffd3d636dcb7bb4b67a3fe9ee858`:
> 
> 1. `cargo test` runs successfully
> 2. `cargo clippy --all-targets` fails
> 
> The problem:
> 
> * Error occurs in: `use crate::common::{execute_query, fetch_results, print_utils};`
> * Clippy searches for the `common` module in the `pg_analytics` crate instead of the `tests` folder
> 
> I tried several solutions (e.g., using `super`) to fix the module path. But no luck. As a temporary fix, I've:
> 
> * Moved the module `fixture` and `common` out of the `tests` folder
> * Created a new crate: `pg_analytics_test_helpers`
> 
> image

This code is added by your PR right? Can we instead fix the import path and keep things in the same crate?

**Comment by shamb0** [comment]
Hi @philippemnoel,

The PR is ready for the next level of review. I've removed the duplicate code, performed a cleanup, and ensured proper integration with PR #91. Additionally, `fixtures/tables/auto_sales.rs` now follows the pattern used in other test implementations across the workspace.

**Comment by philippemnoel** [comment]
Hi @shamb0. This looks super clean! Thank you for integrating everything properly, I'm *very* excited about this PR.

I believe it should also have documentation, so that the users can know that partitions are supported and know how to use them. Our documentation is stored in https://github.com/paradedb/paradedb/tree/dev/docs. Would you be willing to submit a PR with documentation to that repository as well? Then I think everything will be complete :). I'll let Ming do a more thorough review

**Comment by shamb0** [comment]
Hi @philippemnoel,

I wanted to update you on [PR#1568](https://github.com/paradedb/paradedb/pull/1568), which includes recent documentation changes.

Currently, I’ve placed the new topic under `ingest/configuration/multi-level-partitioned-tables`. Could you please review and suggest if this is the most appropriate location, or if you have any preferred path for this topic?

image

**Review by philippemnoel — COMMENTED** [review]
Hi @shamb0! Thank you for doing this. We appreciate the docs PRs. We'll probably push a commit to it to make it more similar to our existing documentation, and then this should be good. I'll let @rebasedming do a proper review

**Comment by shamb0** [comment]
Hi @philippemnoel,

The rebase is complete, and the PR should now be ready for intake review. Please let me know if you encounter any issues.

Thanks again!

**Comment by philippemnoel** [comment]
> Hi @philippemnoel,
> 
> The rebase is complete, and the PR should now be ready for intake review. Please let me know if you encounter any issues.
> 
> Thanks again!

Thank you! Could you please take a look at the failing test?

**Comment by rebasedming** [comment]
Hi @shamb0 , I've spent some time testing this PR and I have some bad news.

While the strategy you used of setting partitions to foreign tables works, it comes at a significant performance penalty. In `pg_analytics`, we push down the entire query to DuckDB by intercepting it in the executor hook. By querying the top-level partitioned table you put in front of the foreign tables, the executor hook is not run, and the query is executed by the Postgres FDW API, which essentially performs a sequential scan of the underlying Parquet file (with predicate/limit pushdown). This scan is significantly slower than a full DuckDB query for lots of use cases and obviates the performance benefits of only scanning one Parquet file.

**Comment by shamb0** [comment]
> Hi @shamb0 , I've spent some time testing this PR and I have some bad news.
> 
> While the strategy you used of setting partitions to foreign tables works, it comes at a significant performance penalty. In `pg_analytics`, we push down the entire query to DuckDB by intercepting it in the executor hook. By querying the top-level partitioned table you put in front of the foreign tables, the executor hook is not run, and the query is executed by the Postgres FDW API, which essentially performs a sequential scan of the underlying Parquet file (with predicate/limit pushdown). This scan is significantly slower than a full DuckDB query for lots of use cases and obviates the performance benefits of only scanning one Parquet file.

Hi @rebasedming,

Thank you for the insightful feedback and for highlighting the performance issue with the strategy used in the PR.

Based on your comments, I understand that the current approach of using partitioned heap tables in front of foreign tables bypasses the executor hook, which normally pushes queries to DuckDB. This results in slower query execution through PostgreSQL's FDW API, as it ends up performing a sequential scan of the underlying Parquet files.

I will investigate this issue further and work on an improved strategy that maintains the performance benefits of DuckDB. I’ll get back to you soon with a better solution.

Thanks for your patience!

**Comment by philippemnoel** [comment]
> > Hi @shamb0 , I've spent some time testing this PR and I have some bad news.
> > While the strategy you used of setting partitions to foreign tables works, it comes at a significant performance penalty. In `pg_analytics`, we push down the entire query to DuckDB by intercepting it in the executor hook. By querying the top-level partitioned table you put in front of the foreign tables, the executor hook is not run, and the query is executed by the Postgres FDW API, which essentially performs a sequential scan of the underlying Parquet file (with predicate/limit pushdown). This scan is significantly slower than a full DuckDB query for lots of use cases and obviates the performance benefits of only scanning one Parquet file.
> 
> Hi @rebasedming,
> 
> Thank you for the insightful feedback and for highlighting the performance issue with the strategy used in the PR.
> 
> Based on your comments, I understand that the current approach of using partitioned heap tables in front of foreign tables bypasses the executor hook, which normally pushes queries to DuckDB. This results in slower query execution through PostgreSQL's FDW API, as it ends up performing a sequential scan of the underlying Parquet files.
> 
> I will investigate this issue further and work on an improved strategy that maintains the performance benefits of DuckDB. I’ll get back to you soon with a better solution.
> 
> Thanks for your patience!

We're excited to see the next iteration :)

**Comment by rebasedming** [comment]
Closing this now as per the above discussion

**Pull request closed without merging**